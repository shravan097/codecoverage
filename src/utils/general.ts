export function filterCoverageByFile(coverage: CoverageParsed): CoverageFile[] {
  return coverage.map(item => ({
    fileName: item.file,
    missingLineNumbers: item?.lines?.details
      .filter(line => line.hit === 0)
      .map(line => line.line)
  }))
}

export function coalesceLineNumbers(lines: number[]): LineRange[] {
  const ranges: LineRange[] = []
  let rstart
  let rend
  for (let i = 0; i < lines.length; i++) {
    rstart = lines[i]
    rend = rstart
    while (lines[i + 1] - lines[i] === 1) {
      rend = lines[i + 1]
      i++
    }
    ranges.push({start_line: rstart, end_line: rend})
  }
  return ranges
}

export function intersectLineRanges(
  a: LineRange[],
  b: LineRange[]
): LineRange[] {
  const result: LineRange[] = []
  let i = 0
  let j = 0

  while (i < a.length && j < b.length) {
    const rangeA = a[i]
    const rangeB = b[j]

    if (rangeA.end_line < rangeB.start_line) {
      i++
    } else if (rangeB.end_line < rangeA.start_line) {
      j++
    } else {
      const start = Math.max(rangeA.start_line, rangeB.start_line)
      const end = Math.min(rangeA.end_line, rangeB.end_line)
      result.push({start_line: start, end_line: end})

      if (rangeA.end_line < rangeB.end_line) {
        i++
      } else {
        j++
      }
    }
  }

  return result
}

export type CoverageParsed = {
  file: string
  title: string
  lines: {
    found: number
    hit: number
    details: {
      line: number
      hit: number
      name: string
    }[]
  }
}[]

export type CoverageFile = {
  fileName: string
  missingLineNumbers: number[]
}

export type LineRange = {
  start_line: number
  end_line: number
}
