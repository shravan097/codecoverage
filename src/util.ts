import * as NodeUtil from 'util'
import * as core from '@actions/core'
import * as fs from 'fs'
import * as github from '@actions/github'
import {Octokit} from 'octokit'
import parser from 'lcov-parse'

export type CoverageFile = {
  fileName: string
  missingLineNumbers: number[]
}

export type LCovParsed = {
  file: string
  title: string
  lines: {
    found: number
    hit: number
    details: {
      line: number
      hit: number
      name: string
    }[]
  }
}[]

export async function parseLCov(lcovPath: string): Promise<LCovParsed> {
  if (!lcovPath) {
    throw Error('No LCov path provided')
  }
  const parserPromise = NodeUtil.promisify(parser)
  const fileRaw = fs.readFileSync(lcovPath, 'utf8')
  return parserPromise(fileRaw) as LCovParsed
}

export function filterCoverageByFile(coverage: LCovParsed): CoverageFile[] {
  return coverage.map(item => ({
    fileName: item.file,
    missingLineNumbers: item?.lines?.details
      .filter(line => line.hit === 0)
      .map(line => line.line)
  }))
}
/** input: dist/index.test.ts --> output: index */
function getFileNameFirstItemFromPath(path: string): string | undefined {
  const rawFileName = path?.split('/')?.pop()
  return rawFileName?.split('.')?.[0]
}

/**
 * https://docs.github.com/en/rest/reference/pulls#list-pull-requests-files
 * Todo update types
 *  */
async function getPullRequestFiles(
  octokitClient: Octokit
): Promise<Set<string>> {
  const pull_number = github.context.issue.number
  const response = await octokitClient.rest.pulls.listFiles({
    ...github.context.repo,
    pull_number
  })
  core.info(`Pull Request Files Length: ${response.data.length}`)
  const mySet = new Set<string>()
  for (const item of response.data) {
    const fileNameFirstItem = getFileNameFirstItemFromPath(item?.filename)
    if (fileNameFirstItem) mySet.add(fileNameFirstItem)
  }
  core.info(`Filename as a set ${mySet.size}`)
  core.info(JSON.stringify(Array.from(mySet)?.splice(0, 10)))
  return mySet
}
type Annotations = {
  path: string,
  start_line: number,
  end_line: number,
  start_column: number,
  end_column: number,
  annotation_level: string,
  message: string
}
export async function annotateGithub(
  coverageFiles: CoverageFile[],
  githubToken: string
): Promise<unknown> {
  if (!githubToken) {
    throw Error('GITHUB_TOKEN is missing')
  }
  const pullRequest = github.context.payload.pull_request
  core.info(`Coverage files length ${coverageFiles?.length}`)
  core.info(JSON.stringify(coverageFiles?.splice(0, 10)))
  core.info(`Pull request number ${github.context.issue.number}`)
  const ref = pullRequest
    ? pullRequest.head.ref
    : github.context.ref.replace('refs/heads/', '')

  const octokit = new Octokit({auth: githubToken})
  const pullRequestFiles = await getPullRequestFiles(octokit)
  const logs: Object[] = []
  const annotations: Annotations[] = []
  for (const current of coverageFiles) {
    // Only annotate relevant files
    const log = {}
    const fileNameFirstItem = getFileNameFirstItemFromPath(current?.fileName)
    log['fileName'] = fileNameFirstItem
    if (fileNameFirstItem && pullRequestFiles.has(fileNameFirstItem)) {
      current.missingLineNumbers.map(lineNumber => {
        annotations.push({
          path: current.fileName,
          start_line: lineNumber,
          end_line: lineNumber,
          start_column: 1,
          end_column: 1,
          annotation_level: 'warning',
          message: 'this line is not covered by test'
        })
      })
      log['missingNumbers'] = current.missingLineNumbers.length
      log['annotationsSize'] = annotations.length
    }
    logs.push(log)
  }
  core.info(JSON.stringify(logs))
  core.info(`Annotation count: ${annotations.length}`)
  core.info(JSON.stringify(annotations.splice(0, 5)))
  const response = await octokit.rest.checks.create({
    ...github.context.repo,
    name: 'Annotate',
    head_sha: ref,
    status: 'completed',
    conclusion: 'success',
    output: {
      title: 'Coverage Tool',
      summary: 'Missing Coverage',
      annotations
    }
  })
  core.info(response.data.output.annotations_url)
  return response
}
