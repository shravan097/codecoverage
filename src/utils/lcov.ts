import * as NodeUtil from 'util'
import * as fs from 'fs'
import parser from 'lcov-parse'

export async function parseLCov(lcovPath: string): Promise<LCovParsed> {
  if (!lcovPath) {
    throw Error('No LCov path provided')
  }
  const parserPromise = NodeUtil.promisify(parser)
  const fileRaw = fs.readFileSync(lcovPath, 'utf8')
  return parserPromise(fileRaw) as LCovParsed
}

export function filterCoverageByFile(coverage: LCovParsed): CoverageFile[] {
  return coverage.map(item => ({
    fileName: item.file,
    missingLineNumbers: item?.lines?.details
      .filter(line => line.hit === 0)
      .map(line => line.line)
  }))
}

export type LCovParsed = {
  file: string
  title: string
  lines: {
    found: number
    hit: number
    details: {
      line: number
      hit: number
      name: string
    }[]
  }
}[]

export type CoverageFile = {
  fileName: string
  missingLineNumbers: number[]
}
