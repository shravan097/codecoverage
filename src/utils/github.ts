import * as core from '@actions/core'
import * as github from '@actions/github'
import {CoverageFile} from './lcov'
import {Octokit} from 'octokit'
import {getFileNameFirstItemFromPath} from './general'

export class Github {
  private client: Octokit

  constructor(token: string) {
    if (!token) {
      throw new Error('GITHUB_TOKEN is missing')
    }
    this.client = new Octokit({auth: token})
  }

  getPullRequestRef(): string {
    const pullRequest = github.context.payload.pull_request
    return pullRequest
      ? pullRequest.head.ref
      : github.context.ref.replace('refs/heads/', '')
  }
  /**
   * https://docs.github.com/en/rest/reference/pulls#list-pull-requests-files
   * Todo update types
   *  */
  async getPullRequestFiles(): Promise<Set<string>> {
    const pull_number = github.context.issue.number
    const response = await this.client.rest.pulls.listFiles({
      ...github.context.repo,
      pull_number
    })
    core.info(`Pull Request Files Length: ${response.data.length}`)
    const mySet = new Set<string>()
    for (const item of response.data) {
      const fileNameFirstItem = getFileNameFirstItemFromPath(item?.filename)
      if (fileNameFirstItem) mySet.add(fileNameFirstItem)
    }
    core.info(`Filename as a set ${mySet.size}`)
    return mySet
  }
  async annotate(input: InputAnnotateParams): Promise<number> {
    // Todo: make this generic
    const response = await this.client.rest.checks.create({
      ...github.context.repo,
      name: 'Annotate',
      head_sha: input.referenceCommitHash,
      status: 'completed',
      conclusion: 'success',
      output: {
        title: 'Coverage Tool',
        summary: 'Missing Coverage',
        annotations: input.annotations
      }
    })
    core.info(response.data.output.annotations_url)
    return response.status
  }

  buildAnnotations(
    coverageFiles: CoverageFile[],
    pullRequestFiles: Set<string>
  ): Annotations[] {
    const annotations: Annotations[] = []
    for (const current of coverageFiles) {
      // Only annotate relevant files
      const fileNameFirstItem = getFileNameFirstItemFromPath(current?.fileName)
      if (fileNameFirstItem && pullRequestFiles.has(fileNameFirstItem)) {
        current.missingLineNumbers.map(lineNumber => {
          annotations.push({
            path: current.fileName,
            start_line: lineNumber,
            end_line: lineNumber,
            start_column: 1,
            end_column: 1,
            annotation_level: 'warning',
            message: 'this line is not covered by test'
          })
        })
      }
    }
    core.info(`Annotation count: ${annotations.length}`)
    return annotations
  }
}

type InputAnnotateParams = {
  referenceCommitHash: string
  annotations: Annotations[]
}
type Annotations = {
  path: string
  start_line: number
  end_line: number
  start_column: number
  end_column: number
  annotation_level: string
  message: string
}
