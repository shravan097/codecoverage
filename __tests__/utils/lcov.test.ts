import {test} from '@jest/globals'
import { parseLCov } from '../../src/utils/lcov'
import {getFixturePath} from '../fixtures/util'

test.only('should parse lCov file', async function () {
  const path = getFixturePath('lcov.info')
  const output = await parseLCov(path)
  expect(output).toMatchSnapshot()
})

xtest('should throw err if lCov file path is not given', async function () {
  await expect(parseLCov('')).rejects.toThrow('No LCov path provided')
})

xtest('filterCoverageByFile', function () {

})

