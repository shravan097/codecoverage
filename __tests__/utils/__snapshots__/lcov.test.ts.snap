// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should parse lCov file 1`] = `
Array [
  Object {
    "branches": Object {
      "details": Array [
        Object {
          "block": 0,
          "branch": 0,
          "line": 3,
          "taken": 0,
        },
        Object {
          "block": 0,
          "branch": 1,
          "line": 3,
          "taken": 0,
        },
        Object {
          "block": 1,
          "branch": 0,
          "line": 3,
          "taken": 0,
        },
        Object {
          "block": 1,
          "branch": 1,
          "line": 3,
          "taken": 0,
        },
        Object {
          "block": 2,
          "branch": 0,
          "line": 3,
          "taken": 0,
        },
        Object {
          "block": 2,
          "branch": 1,
          "line": 3,
          "taken": 0,
        },
        Object {
          "block": 3,
          "branch": 0,
          "line": 3,
          "taken": 0,
        },
        Object {
          "block": 3,
          "branch": 1,
          "line": 3,
          "taken": 0,
        },
        Object {
          "block": 4,
          "branch": 0,
          "line": 4,
          "taken": 0,
        },
        Object {
          "block": 4,
          "branch": 1,
          "line": 4,
          "taken": 0,
        },
        Object {
          "block": 5,
          "branch": 0,
          "line": 4,
          "taken": 0,
        },
        Object {
          "block": 5,
          "branch": 1,
          "line": 4,
          "taken": 0,
        },
        Object {
          "block": 6,
          "branch": 0,
          "line": 4,
          "taken": 0,
        },
        Object {
          "block": 6,
          "branch": 1,
          "line": 4,
          "taken": 0,
        },
        Object {
          "block": 7,
          "branch": 0,
          "line": 4,
          "taken": 0,
        },
        Object {
          "block": 7,
          "branch": 1,
          "line": 4,
          "taken": 0,
        },
      ],
      "found": 16,
      "hit": 0,
    },
    "file": "src/utils/general.ts",
    "functions": Object {
      "details": Array [
        Object {
          "hit": 0,
          "line": 2,
          "name": "getFileNameFirstItemFromPath",
        },
      ],
      "found": 1,
      "hit": 0,
    },
    "lines": Object {
      "details": Array [
        Object {
          "hit": 1,
          "line": 2,
        },
        Object {
          "hit": 0,
          "line": 3,
        },
        Object {
          "hit": 0,
          "line": 4,
        },
      ],
      "found": 3,
      "hit": 1,
    },
    "title": "",
  },
  Object {
    "branches": Object {
      "details": Array [
        Object {
          "block": 0,
          "branch": 0,
          "line": 11,
          "taken": 0,
        },
        Object {
          "block": 1,
          "branch": 0,
          "line": 19,
          "taken": 0,
        },
        Object {
          "block": 1,
          "branch": 1,
          "line": 19,
          "taken": 0,
        },
        Object {
          "block": 2,
          "branch": 0,
          "line": 36,
          "taken": 0,
        },
        Object {
          "block": 2,
          "branch": 1,
          "line": 36,
          "taken": 0,
        },
        Object {
          "block": 3,
          "branch": 0,
          "line": 36,
          "taken": 0,
        },
        Object {
          "block": 3,
          "branch": 1,
          "line": 36,
          "taken": 0,
        },
        Object {
          "block": 4,
          "branch": 0,
          "line": 37,
          "taken": 0,
        },
        Object {
          "block": 5,
          "branch": 0,
          "line": 67,
          "taken": 0,
        },
        Object {
          "block": 5,
          "branch": 1,
          "line": 67,
          "taken": 0,
        },
        Object {
          "block": 6,
          "branch": 0,
          "line": 67,
          "taken": 0,
        },
        Object {
          "block": 6,
          "branch": 1,
          "line": 67,
          "taken": 0,
        },
        Object {
          "block": 7,
          "branch": 0,
          "line": 68,
          "taken": 0,
        },
        Object {
          "block": 8,
          "branch": 0,
          "line": 68,
          "taken": 0,
        },
        Object {
          "block": 8,
          "branch": 1,
          "line": 68,
          "taken": 0,
        },
      ],
      "found": 15,
      "hit": 0,
    },
    "file": "src/utils/github.ts",
    "functions": Object {
      "details": Array [
        Object {
          "hit": 1,
          "line": 10,
          "name": "(anonymous_13)",
        },
        Object {
          "hit": 0,
          "line": 17,
          "name": "(anonymous_14)",
        },
        Object {
          "hit": 0,
          "line": 27,
          "name": "(anonymous_15)",
        },
        Object {
          "hit": 0,
          "line": 42,
          "name": "(anonymous_17)",
        },
        Object {
          "hit": 0,
          "line": 60,
          "name": "(anonymous_19)",
        },
        Object {
          "hit": 0,
          "line": 69,
          "name": "(anonymous_20)",
        },
      ],
      "found": 6,
      "hit": 1,
    },
    "lines": Object {
      "details": Array [
        Object {
          "hit": 1,
          "line": 1,
        },
        Object {
          "hit": 1,
          "line": 2,
        },
        Object {
          "hit": 1,
          "line": 4,
        },
        Object {
          "hit": 1,
          "line": 5,
        },
        Object {
          "hit": 1,
          "line": 7,
        },
        Object {
          "hit": 1,
          "line": 11,
        },
        Object {
          "hit": 0,
          "line": 12,
        },
        Object {
          "hit": 1,
          "line": 14,
        },
        Object {
          "hit": 0,
          "line": 18,
        },
        Object {
          "hit": 0,
          "line": 19,
        },
        Object {
          "hit": 0,
          "line": 28,
        },
        Object {
          "hit": 0,
          "line": 29,
        },
        Object {
          "hit": 0,
          "line": 33,
        },
        Object {
          "hit": 0,
          "line": 34,
        },
        Object {
          "hit": 0,
          "line": 35,
        },
        Object {
          "hit": 0,
          "line": 36,
        },
        Object {
          "hit": 0,
          "line": 37,
        },
        Object {
          "hit": 0,
          "line": 39,
        },
        Object {
          "hit": 0,
          "line": 40,
        },
        Object {
          "hit": 0,
          "line": 44,
        },
        Object {
          "hit": 0,
          "line": 56,
        },
        Object {
          "hit": 0,
          "line": 57,
        },
        Object {
          "hit": 0,
          "line": 64,
        },
        Object {
          "hit": 0,
          "line": 65,
        },
        Object {
          "hit": 0,
          "line": 67,
        },
        Object {
          "hit": 0,
          "line": 68,
        },
        Object {
          "hit": 0,
          "line": 69,
        },
        Object {
          "hit": 0,
          "line": 70,
        },
        Object {
          "hit": 0,
          "line": 82,
        },
        Object {
          "hit": 0,
          "line": 83,
        },
      ],
      "found": 30,
      "hit": 7,
    },
    "title": "",
  },
  Object {
    "branches": Object {
      "details": Array [
        Object {
          "block": 0,
          "branch": 0,
          "line": 6,
          "taken": 0,
        },
        Object {
          "block": 1,
          "branch": 0,
          "line": 17,
          "taken": 0,
        },
        Object {
          "block": 1,
          "branch": 1,
          "line": 17,
          "taken": 4,
        },
        Object {
          "block": 2,
          "branch": 0,
          "line": 17,
          "taken": 4,
        },
        Object {
          "block": 2,
          "branch": 1,
          "line": 17,
          "taken": 4,
        },
        Object {
          "block": 3,
          "branch": 0,
          "line": 17,
          "taken": 0,
        },
        Object {
          "block": 3,
          "branch": 1,
          "line": 17,
          "taken": 4,
        },
        Object {
          "block": 4,
          "branch": 0,
          "line": 17,
          "taken": 4,
        },
        Object {
          "block": 4,
          "branch": 1,
          "line": 17,
          "taken": 4,
        },
      ],
      "found": 9,
      "hit": 6,
    },
    "file": "src/utils/lcov.ts",
    "functions": Object {
      "details": Array [
        Object {
          "hit": 0,
          "line": 5,
          "name": "parseLCov",
        },
        Object {
          "hit": 1,
          "line": 14,
          "name": "filterCoverageByFile",
        },
        Object {
          "hit": 4,
          "line": 15,
          "name": "(anonymous_17)",
        },
        Object {
          "hit": 18,
          "line": 18,
          "name": "(anonymous_18)",
        },
        Object {
          "hit": 12,
          "line": 19,
          "name": "(anonymous_19)",
        },
      ],
      "found": 5,
      "hit": 4,
    },
    "lines": Object {
      "details": Array [
        Object {
          "hit": 1,
          "line": 1,
        },
        Object {
          "hit": 1,
          "line": 2,
        },
        Object {
          "hit": 1,
          "line": 3,
        },
        Object {
          "hit": 1,
          "line": 5,
        },
        Object {
          "hit": 0,
          "line": 6,
        },
        Object {
          "hit": 0,
          "line": 7,
        },
        Object {
          "hit": 0,
          "line": 9,
        },
        Object {
          "hit": 0,
          "line": 10,
        },
        Object {
          "hit": 0,
          "line": 11,
        },
        Object {
          "hit": 1,
          "line": 14,
        },
        Object {
          "hit": 4,
          "line": 15,
        },
        Object {
          "hit": 18,
          "line": 18,
        },
        Object {
          "hit": 12,
          "line": 19,
        },
      ],
      "found": 13,
      "hit": 8,
    },
    "title": "",
  },
]
`;
